### PR TITLE
generalize graph partitioning.

### PR DESCRIPTION
**Pull request by zheng-da**
## Description ##
(Brief description on what this PR is about)

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: src/operator/subgraph/partition_graph.cc [hunk 1]
```diff
@@ -111,8 +100,13 @@ void PrintSubgraph(const std::vector<SimpleNode*>& simple_nodes) {
...
 }
 
+/*
+ * This function traverses the nodes in a computation graph from a starting
+ * node following the input links and output links, and marks all nodes that
+ * can be accessed from the starting node.
+ */
 void LabelSubgraph(const Graph&g,
-                   const std::unordered_set<std::string>& op_names,
+                   SubgraphSelectPtr selectFunc,
```
Comment (reminisce) [inline src/operator/subgraph/partition_graph.cc]: select_func

File: src/operator/subgraph/partition_graph.cc [hunk 2]
```diff
@@ -126,46 +120,57 @@ void LabelSubgraph(const Graph&g,
...
+    if (selectFunc->UseIncomingLink()) {
+      for (auto& e : cur_node->node->inputs) {
+        if (selectFunc->Select(*e.node, subgraph_nodes)) {
+          const auto nid = indexed_graph.node_id(e.node.get());
+          CHECK_LT(nid, simple_nodes.size());
+          // this node has not been visited yet
+          if (simple_nodes[nid]->label == -1) {
+            node_queue.push(simple_nodes[nid].get());
+          } else {
+            CHECK_EQ(simple_nodes[nid]->label, label);
```
Comment (reminisce) [inline src/operator/subgraph/partition_graph.cc]: This disallows adjacent subgraphs. Need to revisit this rule.

File: src/operator/subgraph/partition_graph.cc [hunk 3]
```diff
@@ -126,46 +120,57 @@ void LabelSubgraph(const Graph&g,
...
-// number of subgraphs found
+/*
+ * This function finds subgraphs with all nodes that meet certain criteria.
+ * All nodes in a subgraph are marked with the same label.
+ * All nodes in a subgraph have to be connected with each other. If a node
+ * doesn't meet the given criteria, it will be marked with a separate label.
+ */
 void FindSubgraphs(const Graph& g,
-                   const std::unordered_set<std::string>& op_names,
+                   const SubgraphProperty &subgProperty,
```
Comment (reminisce) [inline src/operator/subgraph/partition_graph.cc]: subgraph_prop

File: src/operator/subgraph/partition_graph.cc [hunk 4]
```diff
@@ -126,46 +120,57 @@ void LabelSubgraph(const Graph&g,
...
                    const std::vector<SimpleNodePtr>& simple_nodes,
                    std::vector<std::vector<SimpleNode*>>* subgraph_nodes) {
   //CHECK(simple_nodes != nullptr);
   const auto& indexed_graph = g.indexed_graph();
   CHECK_EQ(indexed_graph.num_nodes(), simple_nodes.size());
   for (size_t i = 0; i < simple_nodes.size(); ++i) {
     nnvm::Node* node = simple_nodes[i]->node;
-    if (!node->is_variable() && simple_nodes[i]->label == -1 && op_names.count(node->op()->name)) {
+    auto selectFunc = subgProperty.CreateSubgraphSelect();
```
Comment (reminisce) [inline src/operator/subgraph/partition_graph.cc]: select_func

File: src/operator/subgraph/partition_graph.cc [hunk 5]
```diff
@@ -241,6 +245,29 @@ void PrintNodeEntries(const std::vector<nnvm::NodeEntry*>& entries) {
...
+  for (size_t i = 0; i < input_entries.size(); i++) {
+    nnvm::NodeEntry *e = input_entries[i];
+    // If the node is a variable itself, we may want to skip the node.
+    if (e->node->is_variable() && skip_var)
+      continue;
+
+    orig_entries->push_back(*e);
+    nnvm::Symbol sym;
+    sym.outputs.push_back(*e);
+    nnvm::NodePtr n = nnvm::CreateVariableNode(sym.ListOutputNames()[0]);
```
Comment (reminisce) [inline src/operator/subgraph/partition_graph.cc:261]: In order to reorder input entries of the subgraph op node, we need to keep the new variable node name the same as `e->node->attrs.name+to_string(e.index)`. I can change this later.

File: src/operator/subgraph/subgraph_op.cc [hunk 6]
```diff
@@ -27,20 +27,27 @@
 #include "../operator_common.h"
 #include "../elemwise_op_common.h"
 #include "../../imperative/imperative_utils.h"
+#include "./subgraph_op.h"
 
 namespace mxnet {
 namespace op {
 
-struct SubgraphOpState {
+static std::unordered_map<std::string, SubgraphPropertyPtr> properties;
```
Comment (reminisce) [inline src/operator/subgraph/subgraph_op.cc]: It's better to define a singleton class as a wrapper for the global static variable, and the naming is too generic, call it `class SubgraphProps`?

File: src/operator/subgraph/subgraph_op.cc [hunk 7]
```diff
@@ -27,20 +27,27 @@
...
+
+void RegisterSubgraphProperty(SubgraphPropertyPtr property) {
+  properties.insert(std::pair<std::string, SubgraphPropertyPtr>(property->GetType(),
```
Comment (reminisce) [inline src/operator/subgraph/subgraph_op.cc]: Do you need to check the return value of `insert`?

File: src/operator/subgraph/subgraph_op.cc [hunk 8]
```diff
@@ -27,20 +27,27 @@
...
+                                                                property));
+}
+
+class DefaultSubgraphOperator: public SubgraphOperator {
+ public:
   // TODO: initialize uuid
-  SubgraphOpState(const Symbol& sym) :
-      subgraph_sym_(&sym),
+  DefaultSubgraphOperator(const Symbol& sym) : SubgraphOperator(sym),
```
Comment (reminisce) [inline src/operator/subgraph/subgraph_op.cc:47]: Naming: `Operator` -> `Op` for short? Keep consistent with `CachedOp`.

File: src/operator/subgraph/subgraph_op.cc [hunk 9]
```diff
@@ -70,16 +89,74 @@ struct SubgraphOpState {
...
   std::shared_ptr<Executor> subgraph_executor_;
-};  // SubgraphOpState
+};
+
+OpStatePtr SimpleSubgraphProperty::CreateSubgraphOperator(const nnvm::Symbol &sym) const {
+  return OpStatePtr::Create<DefaultSubgraphOperator>(sym);
+}
+
+void DefaultSubgraphOperator::Forward(const OpContext& ctx,
+                                     const std::vector<NDArray>& inputs,
```
Comment (reminisce) [inline src/operator/subgraph/subgraph_op.cc]: alignment

File: src/operator/subgraph/subgraph_op.cc [hunk 10]
```diff
@@ -70,16 +89,74 @@ struct SubgraphOpState {
...
 OpStatePtr CreateSubgraphOpState(const NodeAttrs& attrs,
                                  Context ctx,
                                  const std::vector<TShape>& in_shapes,
                                  const std::vector<int>& in_types) {
   const Symbol& subgraph_sym = nnvm::get<Symbol>(attrs.parsed);
-  return OpStatePtr::Create<SubgraphOpState>(subgraph_sym);
+  auto it = attrs.dict.find("exec_type");
+  CHECK(it != attrs.dict.end());
+  std::string exec_name = it->second;
+  auto it1 = properties.find(exec_name);
```
Comment (reminisce) [inline src/operator/subgraph/subgraph_op.cc]: it1 -> prop_iter

File: src/operator/subgraph/subgraph_op.h [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+  virtual bool UseIncomingLink() const {
+    return true;
+  }
+
+  virtual bool UseOutgoingLink() const {
+    return true;
+  }
+
+  virtual bool Select(const nnvm::Node &n,
+                      const std::vector<sg::SimpleNode*> *subgraph_nodes) {
```
Comment (reminisce) [inline src/operator/subgraph/subgraph_op.h]: I don't think it's a good idea to expose `SimpleNode` struct to users. It is only used to internally ease traversing the computational graph. Let's discuss offline.

File: src/operator/subgraph/subgraph_op.h [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+    return !n.is_variable() && op_names->count(n.op()->name);
+  }
+};
+
+/*
+ * This subgraph property finds a subgraph whose nodes have only operators
+ * within a set. The operators in the subgraph will be executed by _subgraph_op.
+ */
+class SimpleSubgraphProperty: public SubgraphProperty {
+  std::shared_ptr<const std::unordered_set<std::string>> op_names;
```
Comment (reminisce) [inline src/operator/subgraph/subgraph_op.h]: Move this to the end of the class.
https://google.github.io/styleguide/cppguide.html#Declaration_Order

File: src/operator/subgraph/subgraph_op.h [hunk 13]
```diff
@@ -0,0 +1,181 @@
...
+
+}
+
+/*
+ * This provides criteria for selecting nodes in a subgraph.
+ * When a node is passed to this object, the selection criteria may be changed.
+ * We can also specify what links we should use when traversing the neighbor
+ * nodes.
+ */
+class SubgraphSelect {
```
Comment (reminisce) [inline src/operator/subgraph/subgraph_op.h]: SubgraphSelector

File: src/operator/subgraph/subgraph_op.h [hunk 14]
```diff
@@ -0,0 +1,181 @@
...
+ public:
+  virtual ~SubgraphSelect() {
+  }
+  /*
+   * Given a set of nodes that have been selected so far for a subgraph, determine
+   * if the input node should be selected for a subgraph.
+   */
+  virtual bool Select(const nnvm::Node &n,
+                      const std::vector<sg::SimpleNode*> *subgraph_nodes) = 0;
+  virtual bool UseIncomingLink() const = 0;
```
Comment (reminisce) [inline src/operator/subgraph/subgraph_op.h]: UseIncomingEdges

File: src/operator/subgraph/subgraph_op.h [hunk 15]
```diff
@@ -0,0 +1,181 @@
...
+  virtual bool UseOutgoingLink() const = 0;
```
Comment (reminisce) [inline src/operator/subgraph/subgraph_op.h]: UseOutgoingEdges

File: src/operator/subgraph/subgraph_op.h [hunk 16]
```diff
@@ -0,0 +1,181 @@
...
+  virtual void Reset() {
```
Comment (reminisce) [inline src/operator/subgraph/subgraph_op.h]: Pure virtual?

File: src/operator/subgraph/subgraph_op.h [hunk 17]
```diff
@@ -0,0 +1,181 @@
...
+  }
+};
+
+using SubgraphSelectPtr = std::shared_ptr<SubgraphSelect>;
+
+/*
+ * This is the interface of the subgraph operator that executes the computation
+ * in the subgraph.
+ */
+class SubgraphOperator {
```
Comment (reminisce) [inline src/operator/subgraph/subgraph_op.h:74]: SubgraphOp

File: src/operator/subgraph/subgraph_op.h [hunk 18]
```diff
@@ -0,0 +1,181 @@
...
+
+/*
+ * This provides a set of properties for partitioning a graph into subgraphs,
+ * reconstructing a new graph from the subgraphs and creating a subgraph
+ * operator to execute the subgraph.
+ */
+class SubgraphProperty {
+ public:
+  // the criteria of selecting the subgraph nodes.
+  virtual SubgraphSelectPtr CreateSubgraphSelect() const = 0;
```
Comment (reminisce) [inline src/operator/subgraph/subgraph_op.h]: CreateSubgraphSelector

File: src/operator/subgraph/subgraph_op.h [hunk 19]
```diff
@@ -0,0 +1,181 @@
...
+  // create an nnvm node for a given subgraph. Here users can customize how to
+  // execute the operators in the subgraph.
+  virtual nnvm::NodePtr GetSubgraphNode(const nnvm::Symbol &s) const = 0;
```
Comment (reminisce) [inline src/operator/subgraph/subgraph_op.h]: CreateSubgraphNode

File: src/operator/subgraph/subgraph_op.cc [hunk 20]
```diff
@@ -27,20 +27,27 @@
...
 
-struct SubgraphOpState {
+static std::unordered_map<std::string, SubgraphPropertyPtr> properties;
+
+void RegisterSubgraphProperty(SubgraphPropertyPtr property) {
+  properties.insert(std::pair<std::string, SubgraphPropertyPtr>(property->GetType(),
+                                                                property));
+}
+
+class DefaultSubgraphOperator: public SubgraphOperator {
```
Comment (reminisce) [inline src/operator/subgraph/subgraph_op.cc]: DefaultSubgraphOp

File: src/operator/subgraph/subgraph_op.h [hunk 21]
```diff
@@ -0,0 +1,181 @@
...
+};
+
+using SubgraphSelectPtr = std::shared_ptr<SubgraphSelect>;
+
+/*
+ * This is the interface of the subgraph operator that executes the computation
+ * in the subgraph.
+ */
+class SubgraphOperator {
+  nnvm::Symbol subgraph_sym_;
```
Comment (reminisce) [inline src/operator/subgraph/subgraph_op.h]: Move this to the end of the class.
https://google.github.io/styleguide/cppguide.html#Declaration_Order

File: src/operator/subgraph/subgraph_op.h [hunk 22]
```diff
@@ -0,0 +1,181 @@
...
+  virtual nnvm::NodePtr GetSubgraphNode(const nnvm::Symbol &s) const = 0;
+  // Create a subgraph operator for execution.
+  virtual OpStatePtr CreateSubgraphOperator(const nnvm::Symbol &sym) const = 0;
+  // The type of the subgraph.
+  virtual std::string GetType() const = 0;
+};
+
+using SubgraphPropertyPtr = std::shared_ptr<SubgraphProperty>;
+
+void RegisterSubgraphProperty(SubgraphPropertyPtr property);
```
Comment (reminisce) [inline src/operator/subgraph/subgraph_op.h:121]: Where is this called?

File: src/operator/subgraph/subgraph_op.h [hunk 23]
```diff
@@ -0,0 +1,181 @@
...
+
+/*
+ * This selects nodes for a subgraph that only contains operators
+ * in a given set and it visits nodes via both input and output links.
+ */
+class ContainOpSelect: public SubgraphSelect {
```
Comment (reminisce) [inline src/operator/subgraph/subgraph_op.h]: Select -> Selector

File: src/operator/subgraph/subgraph_op.h [hunk 24]
```diff
@@ -0,0 +1,181 @@
...
+  std::shared_ptr<const std::unordered_set<std::string>> op_names;
+
+ public:
+  ContainOpSelect(std::shared_ptr<const std::unordered_set<std::string>> op_names) {
```
Comment (reminisce) [inline src/operator/subgraph/subgraph_op.h]: Why `shared_ptr`?

File: src/operator/subgraph/subgraph_op.h [hunk 25]
```diff
@@ -0,0 +1,181 @@
...
+ */
+class SubgraphProperty {
+ public:
+  // the criteria of selecting the subgraph nodes.
+  virtual SubgraphSelectPtr CreateSubgraphSelect() const = 0;
+  // create an nnvm node for a given subgraph. Here users can customize how to
+  // execute the operators in the subgraph.
+  virtual nnvm::NodePtr GetSubgraphNode(const nnvm::Symbol &s) const = 0;
+  // Create a subgraph operator for execution.
+  virtual OpStatePtr CreateSubgraphOperator(const nnvm::Symbol &sym) const = 0;
```
Comment (reminisce) [inline src/operator/subgraph/subgraph_op.h]: CreateSubgraphOp

File: src/operator/subgraph/subgraph_op.h [hunk 26]
```diff
@@ -0,0 +1,181 @@
...
+/*
+ * This subgraph property finds a subgraph whose nodes have only operators
+ * within a set. The operators in the subgraph will be executed by _subgraph_op.
+ */
+class SimpleSubgraphProperty: public SubgraphProperty {
+  std::shared_ptr<const std::unordered_set<std::string>> op_names;
+
+ public:
+  SimpleSubgraphProperty(const std::unordered_set<std::string> &op_names) {
+    this->op_names = std::make_shared<std::unordered_set<std::string>>(op_names);
```
Comment (reminisce) [inline src/operator/subgraph/subgraph_op.h]: Why shared_ptr?

File: src/operator/subgraph/subgraph_op.h [hunk 27]
```diff
@@ -0,0 +1,181 @@
...
+  }
+  virtual nnvm::NodePtr GetSubgraphNode(const nnvm::Symbol &sym) const {
+    nnvm::NodePtr n = nnvm::Node::Create();
+    n->attrs.op = Op::Get("_subgraph_op");
+    n->attrs.name = "_subgraph_op";
+    n->attrs.dict.insert(std::pair<std::string, std::string>("exec_type", GetType()));
+    n->attrs.parsed = std::move(sym);
```
Comment (reminisce) [inline src/operator/subgraph/subgraph_op.h]: Do not move on a const reference.

File: src/operator/subgraph/subgraph_op.h [hunk 28]
```diff
@@ -0,0 +1,181 @@
...
+   */
+  virtual bool Select(const nnvm::Node &n,
+                      const std::vector<sg::SimpleNode*> *subgraph_nodes) = 0;
+  virtual bool UseIncomingLink() const = 0;
+  virtual bool UseOutgoingLink() const = 0;
+  virtual void Reset() {
+  }
+};
+
+using SubgraphSelectPtr = std::shared_ptr<SubgraphSelect>;
```
Comment (reminisce) [inline src/operator/subgraph/subgraph_op.h]: SubgraphSelectorPtr

**Review by reminisce — CHANGES_REQUESTED** [review]
Have you run the unit test in test_subgraph_op.py?

File: src/operator/subgraph/subgraph_op.h (hunk 11 above)
Comment (zheng-da) [reply src/operator/subgraph/subgraph_op.h]: I agree.

File: src/operator/subgraph/subgraph_op.h (hunk 22 above)
Comment (zheng-da) [reply src/operator/subgraph/subgraph_op.h:121]: ideally, there should be something like NNVM_REGISTER_OP, so users can register directly. do you know how to do that?

Comment (reminisce) [reply src/operator/subgraph/subgraph_op.h:121]: I think we can use `dmlc::Registry<SubgraphProperty>`, similar to `dmlc::Registry<::nnvm::Op>` used in `NNVM_REGISTER_OP`.

File: src/operator/subgraph/partition_graph.cc [hunk 29]
```diff
@@ -283,23 +304,40 @@ Graph PartitionGraph(Graph&& g) {
     return ret;
   } else {
     using namespace sg;
+    SubgraphPropertyPtr subg_prop = g.GetAttr<SubgraphPropertyPtr>("subgraph_property");
```
Comment (reminisce) [inline src/operator/subgraph/partition_graph.cc:307]: This should be `const SubgraphProperty&`.

File: src/operator/subgraph/subgraph_op.cc [hunk 30]
```diff
@@ -70,16 +91,88 @@ struct SubgraphOpState {
...
                                  const std::vector<int>& in_types) {
   const Symbol& subgraph_sym = nnvm::get<Symbol>(attrs.parsed);
-  return OpStatePtr::Create<SubgraphOpState>(subgraph_sym);
+  auto it = attrs.dict.find("exec_type");
+  if (it == attrs.dict.end()) {
+    auto op = std::make_shared<DefaultSubgraphOperator>(subgraph_sym);
+    return OpStatePtr::Create<SubgraphOpState>(op);
+  }
+
+  std::string exec_name = it->second;
```
Comment (reminisce) [inline src/operator/subgraph/subgraph_op.cc:170]: const std::string

File: src/operator/subgraph/subgraph_op.cc [hunk 31]
```diff
@@ -70,16 +91,88 @@ struct SubgraphOpState {
...
+
+SubgraphOperatorPtr SimpleSubgraphProperty::CreateSubgraphOperator(const nnvm::Symbol &sym) const {
+  return std::make_shared<DefaultSubgraphOperator>(sym);
+}
+
+void DefaultSubgraphOperator::Forward(const OpContext& ctx,
+                                      const std::vector<NDArray>& inputs,
+                                      const std::vector<OpReqType>& req,
+                                      const std::vector<NDArray>& outputs) {
+  // We can create an executor to run this subgraph op
```
Comment (reminisce) [inline src/operator/subgraph/subgraph_op.cc:105]: I will change the implementation here to adopt CachedOp since we want to support varied batch size seamlessly.
